File: SortWare/DupeStruct.vb
Public Class DupeStruct

    Private Class FileNamePair
        Private Dir1Files As List(Of String)
        Private Dir2Files As List(Of String)

        Public Sub New()
            Dir1Files = New List(Of String)
            Dir2Files = New List(Of String)
        End Sub

        Public Sub New(ByVal d1f As String, ByVal d2f As String)
            Dir1Files = New List(Of String)
            Dir2Files = New List(Of String)

            If d1f IsNot Nothing Then
                Dir1Files.Add(d1f)
            End If

            If d2f IsNot Nothing Then
                Dir2Files.Add(d2f)
            End If
        End Sub

        Public Sub AddD1File(ByVal fname As String)
            Dir1Files.Add(fname)
        End Sub

        Public Sub AddD2File(ByVal fname As String)
            Dir2Files.Add(fname)
        End Sub

        Public Function Count() As Long
            Return Dir1Files.Count + Dir2Files.Count
        End Function

        Public Function isGoodCount() As Boolean
            Return (Count() > 1)
        End Function

        Public Function GetDir1Files() As List(Of String)
            Return New List(Of String)(Dir1Files)
        End Function

        Public Function GetDir2Files() As List(Of String)
            Return New List(Of String)(Dir2Files)
        End Function
    End Class

    Private dupeDict As Dictionary(Of Long, FileNamePair)

    Public Sub New()
        dupeDict = New Dictionary(Of Long, FileNamePair)
    End Sub

    Public Sub AddD1File(ByVal size As Long, ByVal fname As String)
        If Not dupeDict.ContainsKey(size) Then
            dupeDict.Add(size, New FileNamePair(fname, Nothing))
        Else
            dupeDict.Item(size).AddD1File(fname)
        End If
    End Sub

    Public Sub AddD2File(ByVal size As Long, ByVal fname As String)
        If Not dupeDict.ContainsKey(size) Then
            dupeDict.Add(size, New FileNamePair(Nothing, fname))
        Else
            dupeDict.Item(size).AddD2File(fname)
        End If
    End Sub

    Public Sub ClearLowCountSizes()
        For Each item In dupeDict.ToList
            If Not item.Value.isGoodCount Then
                dupeDict.Remove(item.Key)
            End If
        Next
    End Sub

    Public Function GetDir1Files() As List(Of String)
        Dim ret As List(Of String) = New List(Of String)
        For Each item In dupeDict
            ret.AddRange(item.Value.GetDir1Files)
        Next
        Return ret
    End Function

    Public Function GetDir2Files() As List(Of String)
        Dim ret As List(Of String) = New List(Of String)
        For Each item In dupeDict
            ret.AddRange(item.Value.GetDir2Files)
        Next
        Return ret
    End Function

    Public Sub Clear()
        dupeDict.Clear()
    End Sub
End Class

File: SortWare/DupeStruct.cs
using System.Collections.Generic;
using System.Linq;

namespace SortWare
{
    public class DupeStruct
    {

        private class FileNamePair
        {
            private List<string> Dir1Files;
            private List<string> Dir2Files;

            public FileNamePair()
            {
                Dir1Files = new List<string>();
                Dir2Files = new List<string>();
            }

            public FileNamePair(string d1f, string d2f)
            {
                Dir1Files = new List<string>();
                Dir2Files = new List<string>();

                if (d1f is not null)
                {
                    Dir1Files.Add(d1f);
                }

                if (d2f is not null)
                {
                    Dir2Files.Add(d2f);
                }
            }

            public void AddD1File(string fname)
            {
                Dir1Files.Add(fname);
            }

            public void AddD2File(string fname)
            {
                Dir2Files.Add(fname);
            }

            public long Count()
            {
                return Dir1Files.Count + Dir2Files.Count;
            }

            public bool isGoodCount()
            {
                return Count() > 1L;
            }

            public List<string> GetDir1Files()
            {
                return new List<string>(Dir1Files);
            }

            public List<string> GetDir2Files()
            {
                return new List<string>(Dir2Files);
            }
        }

        private Dictionary<long, FileNamePair> dupeDict;

        public DupeStruct()
        {
            dupeDict = new Dictionary<long, FileNamePair>();
        }

        public void AddD1File(long size, string fname)
        {
            if (!dupeDict.ContainsKey(size))
            {
                dupeDict.Add(size, new FileNamePair(fname, null));
            }
            else
            {
                dupeDict[size].AddD1File(fname);
            }
        }

        public void AddD2File(long size, string fname)
        {
            if (!dupeDict.ContainsKey(size))
            {
                dupeDict.Add(size, new FileNamePair(null, fname));
            }
            else
            {
                dupeDict[size].AddD2File(fname);
            }
        }

        public void ClearLowCountSizes()
        {
            foreach (var item in dupeDict.ToList())
            {
                if (!item.Value.isGoodCount())
                {
                    dupeDict.Remove(item.Key);
                }
            }
        }

        public List<string> GetDir1Files()
        {
            var ret = new List<string>();
            foreach (var item in dupeDict)
                ret.AddRange(item.Value.GetDir1Files());
            return ret;
        }

        public List<string> GetDir2Files()
        {
            var ret = new List<string>();
            foreach (var item in dupeDict)
                ret.AddRange(item.Value.GetDir2Files());
            return ret;
        }

        public void Clear()
        {
            dupeDict.Clear();
        }
    }
}